### PR TITLE
In app search improvements feature branch

### PR DESCRIPTION
**Pull request by bor0**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR is based on a feature branch that covers #51266, #51297, #51309, #51313, #51342, #51434.

We needed this feature branch to address issue dependencies.

### How to test the changes in this Pull Request:

1. Open each issue and test it
2. Make sure the search works, do some testing around it, and see if you can spot weird behavior

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Comment by andfinally** [comment]
Just testing https://github.com/woocommerce/woocommerce/pull/51266, and I'm only getting the count bubbles when I do a search.

I wonder if we need to announce `0` results? What do we think @poligilad-auto? Do we even need a bubble when there are no results? (The `0` on My subscriptions is also weird, but that must be a separate issue.)

<p>
image
</p>

**Comment by andfinally** [comment]
https://github.com/woocommerce/woocommerce/pull/51297

* I get a count of 60 for Extensions and 5 for Themes from a search for "Bookings". `http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions&term=Bookings`. 
* The [endpoint](https://woocommerce.com/wp-json/wccom-extensions/1.0/search?term=Bookings&country=gb) returns a total products count of 63 for my country `GB`. All of the items returned have `"type": "extension",` so I'm assuming there are no themes in those results.
* Note for my future self. We've integrated categories into the bubble count. The category only applies within the selected tab, and is usually unique to it. So when you select a tab, the bubble on the other tabs typically goes to 0.

Nothing else to quibble about with https://github.com/woocommerce/woocommerce/pull/51297!

**Comment by KokkieH** [comment]
> I get a count of 60 for Extensions
> The [endpoint](https://woocommerce.com/wp-json/wccom-extensions/1.0/search?term=Bookings&country=gb) returns a total products count of 63 for my country GB

The count only looks at the number of products in the filtered product list from the allProducts value, so the count is based on the number of products currently displayed, not the number of products available on WCCOM. But as the API now provides that value we can perhaps update how this is calculated?

**Comment by bor0** [comment]
Yes, after https://github.com/Automattic/woocommerce.com/pull/21678/files it seems like a good idea to rewrite how `setSearchResultsCount` is used. (Another dependency-issue we didn't consider 🙂)

**Comment by andfinally** [comment]
Ah right, I get it. Yes, I'd update that, but it could wait to next release. I just wanted to spell out my results in case there was an issue.

**Comment by bor0** [comment]
@andfinally I addressed that in 3223370. There is one small edge-case which, given the current code and architecture might be a bit tricky to work around.

To reproduce:

1. Go to http://localhost:8080/wp-admin/admin.php?page=wc-admin&path=%2Fextensions
2. Click "Extensions"
3. Search "Woo". Counts show as 808, 91, 5.
4. Select "Developed by Woo". Counts show 61, 91, 5.
5. Search for "Wooo". Counts show 0, 91, 5 instead of 0, 0, 0, since we're not doing extra fetches for the other categories, when there is a category selected.
    - I thought it's better to leave the previous number rather than reset it to 0 (we can do that easily by doing `setSearchResultsCount( { extensions: 0, themes: 0, 'business-services': 0, [ category ]: count } )`

Addressing this might also slightly impact performance.

**Comment by andfinally** [comment]
https://github.com/woocommerce/woocommerce/pull/51266 is good. I get slight differences in some of my counts, but that is probably due to country.

https://github.com/woocommerce/woocommerce/pull/51309 is good.

While testing this I noticed the themes cards overflow on mobile view. Added an [issue](21725-gh-Automattic/woocommerce.com) for this we can look at another time.

**Comment by KokkieH** [comment]
> While testing this I noticed the themes cards overflow on mobile view

Hmm, @raicem fixed that the other day. I wonder if a regression slipped in when the category selector PR was rebased on the feature branch.

**Comment by KokkieH** [comment]
@andfinally the theme card issue was fixed in https://github.com/woocommerce/woocommerce/pull/51187, which is already in trunk. If you update this branch against trunk you shouldn't see the issue any more.

**Comment by andfinally** [comment]
https://github.com/woocommerce/woocommerce/pull/51313 is good. Thanks for thinking about accessibility! I noticed some unrelated accessibility issues – I created an [issue](21727-gh-Automattic/woocommerce.com), we can add to it as we find them.

https://github.com/woocommerce/woocommerce/pull/51342 looks good. I don't get the header on the "no results" message on the video on that PR. I also get the category tabs. But I guess that is a decision made during development.

<p>
image
</p>

> @andfinally the theme card issue was fixed in https://github.com/woocommerce/woocommerce/pull/51187, which is already in trunk. If you update this branch against trunk you shouldn't see the issue any more.

Awesome, it's sorted! Closing my issue for this.

https://github.com/woocommerce/woocommerce/pull/51434 is good as well. Lovely job!

I still have to look through the code, especially https://github.com/woocommerce/woocommerce/commit/3223370315842fbd2dcd8d7da7c83c2fa4795f2c, but functionality-wise, we're looking good.

**Review by KokkieH — COMMENTED** [review]
On the Discover tab, if I do a search it switches me to the Extensions tab. However, if I do a search from the Subscriptions tab, it doesn't switch me. Is this intentional?

When searching, there's no screen reader announcement that the page content has been updated. It would be good if we can improve this before merging. [Here's a patch](https://github.com/user-attachments/files/17039719/search-a11y.patch) with a possible fix for this--doesn't work for the category filters, but it does work for search both with and without category filters set. With a category filter set, it should only announce the count for the current tab, else it announces the total across all tabs.

Aside from that everything tests well.

**Comment by raicem** [comment]
> Hmm, @raicem fixed that the other day. I wonder if a regression slipped in when the category selector PR was rebased on the feature branch.

Thank you for catching this @andfinally and @KokkieH. It's indeed deleted now from this branch. I'll push a commit to add it back now. The header is also not sticky, so we lost that too. I wonder what I did wrong. I [merged](https://github.com/woocommerce/woocommerce/pull/51187) the sticky header directly to the `trunk`. 

edit: Herman warned me the branch was just behind on trunk, so no action needed

**Comment by poligilad-auto** [comment]
Just jumping in here to say, please just notice that the [bubbles with search results](https://www.figma.com/design/2sXEw33oLyTZ7Mu3808sed/In-App-Marketplace?node-id=3916-72596&t=qWwmSnGMPYPW7sdO-1) should have different colors.

- Selected tab is black.
- Other tabs is gray.

The designs have a "0" as well, that's not a must, but I think the tab should be clickable and show the "No results" page.

**Comment by andfinally** [comment]
> On the Discover tab, if I do a search it switches me to the Extensions tab. However, if I do a search from the Subscriptions tab, it doesn't switch me. Is this intentional?

This is maybe a consequence of the fact that the search doesn't cover My Subscriptions. Now that you mention it, it's not consistent. We can address it in a later release, if we don't get a chance to do it now.

> The designs have a "0" as well, that's not a must

I'm getting a red 0 bubble on My Subscriptions as well. IMHO we shouldn't show a bubble there if there's nothing to announce – the red colour makes it especially distracting. 

I'm wondering if the logic that shows the 0 bubbles on the search tabs is somehow leaking on to the My Subscriptions tab. 

When I delete the WooCommerce Update Manager I "have no subscription for", it changes to a 1 to reflect the notice to install WUM. When I install WUM from a download, it's shown as `No subscription`, and the bubble goes back to 0. It's presumably some issue with how we calculate the count for that particular bubble. Created issue for this for later: 21738-gh-Automattic/woocommerce.com.

<p>image</p>

Personally I'd be in favour of hiding bubbles if we have no count to show. But it's probably not a good idea to make this change with a few hours to go before code freeze.

```diff
Index: plugins/woocommerce-admin/client/marketplace/components/tabs/tabs.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/plugins/woocommerce-admin/client/marketplace/components/tabs/tabs.tsx b/plugins/woocommerce-admin/client/marketplace/components/tabs/tabs.tsx
--- a/plugins/woocommerce-admin/client/marketplace/components/tabs/tabs.tsx	(revision 47870c1a705a8e16e59a3fb2f1fa03ec38fdec8f)
+++ b/plugins/woocommerce-admin/client/marketplace/components/tabs/tabs.tsx	(date 1726653900461)
@@ -106,7 +106,10 @@
 					key={ tabKey }
 				>
 					{ tabs[ tabKey ]?.title }
-					{ tabs[ tabKey ]?.showUpdateCount && (
+					{ 
+						tabs[ tabKey ]?.showUpdateCount
+						&& tabs[ tabKey ]?.updateCount > 0 
+						&& (
 						<span
 							className={ clsx(
 								'woocommerce-marketplace__update-count',
```

<p>image</p>

**Comment by bor0** [comment]
> On the Discover tab, if I do a search it switches me to the Extensions tab. However, if I do a search from the Subscriptions tab, it doesn't switch me

Not intentional, nice catch. Fixed in 2b32972ec952d9788591082ef92c6d8d044bc9b8

> When searching, there's no screen reader announcement that the page content has been updated

This is a nice touch. Applied your patch in 21e675e2b9f5a5e95c77b21168fe16c7a6489da5

> Personally I'd be in favour of hiding bubbles if we have no count to show

This seems like a low risk. Applied in a93b8d8bcd9ce7e8b8d09b17775b6399ca1f962b

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx [hunk 1]
```diff
@@ -52,7 +69,100 @@ export default function Content(): JSX.Element {
...
+						).length,
+						'business-services': combinedProducts.filter(
+							( p ) => p.type === ProductType.businessService
+						).length,
+					} );
+
+					return combinedProducts;
+				} );
+
+				speak( __( 'More products loaded', 'woocommerce' ) );
```
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:135]: Nice!

**Review by andfinally — APPROVED** [review]
Beautiful! I checked the code for the changes from the last couple of days and retested them, all looks good. Thanks for the last-minute tweaks, which make it much better. 🙌

**Comment by jorgemd24** [comment]
Hey @bor0! Thanks for your work on this. I’ve been testing this feature (see here for context pb22l9-3kq-p2 )  and I noticed that clearing the search isn't working as expected. When I click the "X," the text clears, but the filtered results remain, so it's unclear how to return to the full list. I also noticed that the search only works after hitting "Enter," but I’d expect the results to filter as soon as I finish typing. Feel free to let me know if you'd like me to create an issue for this, or if I'm missing something!


https://github.com/user-attachments/assets/505a4fa3-3a07-4467-a9f0-4c089881eca2

**Comment by bor0** [comment]
@jorgemd24 thanks for noticing this. The PR didn't touch that part specifically, so it is a [known "issue"](https://github.com/woocommerce/woocommerce/blob/ce66b55/plugins/woocommerce-admin/client/marketplace/components/search/search.tsx#L58-L64). While it may be tricky to account for that use case*, feel free to create an issue and we can investigate.

*: There is no event we can use when the input resets, we can only rely on `input` or `click`, but we can't simply submit the search when the text is empty because a user can start typing, then change their mind (clear text) then start typing again.

**Comment by jorgemd24** [comment]
Thanks @bor0 for your explanation, I opened this issue: https://github.com/woocommerce/woocommerce/issues/51687

**Pull request merged**